.github/workflows/ci.yml

```diff
@@ -34,6 +34,98 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
+      - name: Setting build env
+        run: |
+          echo "RUST_OS_NAME=${{ runner.os }}" >> $GITHUB_ENV
+        shell: bash
+
+      # We install gnu-tar because BSD tar is buggy on Github's macos machines.
+      # See <https://github.com/actions/cache/issues/403>.
+      - name: Install GNU tar
+        if: runner.os == 'macOS'
+        run: |
+          brew install gnu-tar
+          echo "/usr/local/opt/gnu-tar/libexec/gnubin" >> $GITHUB_PATH
+
+      # Cache the global cargo directory, but NOT the local `target` directory which
+      # we cannot reuse anyway when the nightly changes (and it grows quite large
+      # over time).
+      - name: Add cache for cargo
+        uses: actions/cache@v2
+        with:
+          path: |
+            # to avoid saving rustup proxy, save xargo and RTIM in separate dir.
+            ~/.cargo/ci
+            ~/.cargo/registry
+            ~/.cargo/git
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: ${{ runner.os }}-cargo
+
+      # We use the "stable" toolchain for better caching, it is just used to build `rustup-toolchain-install-master`.
+      # But we also need to take into account that the build cache might have a different, outdated default.
+      - name: Install stable Rust
+        run: |
+          rustup toolchain install stable
+          rustup default stable
+
+      - name: Install RTIM
+        run: |
+          cargo install --root ~/.cargo/ci rustup-toolchain-install-master
```

Comment:
Why `--root ~/.cargo/ci`? This was not needed before.

Reply:
By default, cargo saves CLIs in .cargo/bin dir, but that also store rustup proxies like cargo, rustc.
To optimize cache usage space, we save them in separate dirs.

Reply:
Is this a common optimizations? Sounds like a microoptimization to me; those proxies are all hardlinked (so they shouldn't take up extra space) and even if the caching logic does not understand hardlinks this is just a few magebytes.

How much time does this actually save? I am not convinced it justifies the extra complexity.

Reply:
Another reason to do is that to avoid rebuilding in next CI sessions.
cargo saves files in .cargo/ to preserve the installation state: determine if it reinstalling next time or not.
By saving .cargo/ci, we could save cargo installation info flawlessly.

Reply:
So far we have cached `.cargo/` and we avoid rebuilding these things. So I do not follow what you are saying here. We can just continue to do what we did so far.

You seem to be adding a lot of extra complexity just to save a few megabytes.

Reply:
There is a bug Linux builders of Actions that doesn't cache .cargo at all, they just caches
the hardlink point to another directory.

Reply:
Well then we can list the relevant files and directories in `.cargo`. That's still better because it is an entirely local fix. Also then there should be a `FIXME: use better config once GHA is not broken any more` or so.

(I assume you mean "softlink", there are no directory hardlinks.)

Reply:
Yeah, it's softlink. Honestly, I'm not sure the real problem is. Reading some issues on actions/cache repo
doesn't give me relevant results.

Reply:
How do other Rust projects handle this? If this is a GHA problem, surely we are not the first...

Clippy has
```
    - name: Cache cargo dir
      uses: actions/cache@v2
      with:
        path: ~/.cargo
        key: ${{ runner.os }}-x86_64-unknown-linux-gnu-${{ hashFiles('Cargo.lock') }}
        restore-keys: |
          ${{ runner.os }}-x86_64-unknown-linux-gnu
```
So if it works for them I'd expect it should also work for us? Not sure what `restore-keys` does though.

Reply:
Apparently, caching on Ubuntu builders of clippy caches nothing: <https://github.com/rust-lang/rust-clippy/pull/6154/checks?check_run_id=1235351421#step:6:1>

Reply:
> How do other Rust projects handle this?

[Cargo](https://github.com/rust-lang/cargo/tree/master/.github/workflows) doesn't use actions/cache, they download mdbook from gh release.
[Rustup](https://github.com/rust-lang/rustup/blob/243b453934929ea19f4172f626018531bab45ddd/ci/actions-templates/centos-fmt-clippy-template.yaml#L45) does something similar like this PR.